Sources/IBMSwiftSDKCore/Authentication/IAMAuthenticator.swift

```diff
@@ -70,28 +70,41 @@ class IAMTokenSource: TokenSource {
     }
 
     func getToken(completionHandler: @escaping (String?, RestError?) -> Void) {
-        // request a new access token if the current token is expired
-        guard let refreshDate = refreshDate, refreshDate.timeIntervalSinceNow > 0 else {
-            requestToken {
-                token, error in
-
-                guard let token = token, error == nil else {
-                    completionHandler(nil, error)
-                    return
-                }
-                self.token = token.accessToken
-                // Compute refreshDate for token
-                let expirationDate = Date(timeIntervalSince1970: Double(token.expiration))
-                // We want a little buffer to make sure we refresh proactively
-                let buffer = expirationDate.timeIntervalSinceNow * -0.2
-                self.refreshDate = expirationDate.addingTimeInterval(buffer)
-                completionHandler(self.token, nil)
-            }
+        // If we have a token, pass it to the completion handler.
+        if let token = self.token {
+            completionHandler(token, nil)
+            refreshToken { (_, _) in /* dummy completion handler */}
+        } else {
+            refreshToken(completionHandler: completionHandler)
+        }
+    }
+
+    // request a new access token if the we have passed the refresh time for the current token
+    func refreshToken(completionHandler: @escaping (String?, RestError?) -> Void) {
+        // If refreshDate is set and still in the future, no refresh needed -- just return
+        if let refreshDate = refreshDate, refreshDate.timeIntervalSinceNow > 0 {
             return
         }
 
-        // use the existing, valid access token
-        completionHandler(self.token, nil)
+        // Ideally we'd check and update refreshDate as an atomic operation, but atomic operations are not well
+        // supported in Swift, and the worst thing that can happen here is that we get multiple token requests
+        // issued in the window between the read and update of refreshDate.
+        self.refreshDate = Date(timeIntervalSinceNow: 60) // Update refreshDate to 60 seconds from now
```

Comment:
I guess based on this comment that the window is considered sufficiently small that it isn't worth bringing in GCD or some other locking mechanism. I think that ultimately depends on the levels of concurrency in play and whether the extra requests are sufficient to cause a problem.

I honestly don't know the semantics of access here well enough to estimate the size of this window and how many concurrent threads would be needed to turn it into a problem. I'll see if I can dig up anything on this.

Reply:
I had a bit of play with a facsimile of this approach using a queue of concurrent tasks doing this type of check/set. Basically 72% of invocations result in 1 token request and 28% of them result in 2. I did also see a few outliers with 3 to 5(!) requests.

It wasn't super scientific and was just running in the Swift playground on my laptop, but I guess it gives a rough idea of how frequently this might happen.

Reply:
I could use a serial dispatch queue to serialize the update of the `refreshDate`.  That might not be too onerous.  I'll explore that.